scrypto/src/component/kv_store.rs

```diff
@@ -10,36 +11,76 @@ use scrypto::core::DataAddress;
 
 use crate::abi::*;
 use crate::buffer::*;
+use crate::core::{DataRef, DataRefMut};
 use crate::crypto::*;
 use crate::engine::{api::*, call_engine, types::KeyValueStoreId};
 use crate::misc::*;
+use crate::prelude::HashMap;
 
 /// A scalable key-value map which loads entries on demand.
-#[derive(PartialEq, Eq, Hash)]
-pub struct KeyValueStore<K: Encode + Decode, V: Encode + Decode> {
+pub struct KeyValueStore<K: Encode + Decode, V: 'static + Encode + Decode + TypeId> {
     pub id: KeyValueStoreId,
+    pub map: RefCell<HashMap<DataAddress, Option<V>>>,
     pub key: PhantomData<K>,
     pub value: PhantomData<V>,
 }
 
-impl<K: Encode + Decode, V: Encode + Decode> KeyValueStore<K, V> {
+impl<K: Encode + Decode, V: 'static + Encode + Decode + TypeId> KeyValueStore<K, V> {
     /// Creates a new key value store.
     pub fn new() -> Self {
         let input = RadixEngineInput::CreateKeyValueStore();
         let output: KeyValueStoreId = call_engine(input);
 
         Self {
             id: output,
+            map: RefCell::new(HashMap::new()),
             key: PhantomData,
             value: PhantomData,
         }
     }
 
     /// Returns the value that is associated with the given key.
-    pub fn get(&self, key: &K) -> Option<V> {
+    pub fn get(&self, key: &K) -> Option<DataRef<V>> {
         let address = DataAddress::KeyValueEntry(self.id, scrypto_encode(key));
-        let input = RadixEngineInput::ReadData(address);
-        call_engine(input)
+        let mut borrowed_map = self.map.borrow_mut();
+        if !borrowed_map.contains_key(&address) {
+            let input = ::scrypto::engine::api::RadixEngineInput::ReadData(address.clone());
+            let value: Option<V> = call_engine(input);
+            borrowed_map.insert(address.clone(), value);
+        }
+
+        if borrowed_map.get(&address).unwrap().is_some() {
+            let ref_mut = RefMut::map(borrowed_map, |map| {
+                map.get_mut(&address).unwrap().as_mut().unwrap()
+            });
+
+            Some(DataRef { value: ref_mut })
```

Comment:
This is broken in my view. When the value exists, the `borrow_mut` is never released, so I even be able to do two `get`:
```
{
    let v1 = map.get(&"1".to_owned());
    let v2 = map.get(&"2".to_owned());
}
```

May need to resort to interior mutability and per-entry borrow.

Reply:
Yes, I can add back interior mutability but this would have to go away if we started adding mutable iterators. But I'm good either way.

Reply:
Also there's the problem of protecting against:
```
{
    let v1 = map.get(&"1".to_owned());
    let v2 = map.get(&"1".to_owned());
}
```

Reply:
I'd think the right interface would be more like Rust hashmap `get_many_mut`.

Reply:
Whatever is easier for you. Can we also add a test to cover this scenario?  :)

Reply:
Oh wait, I think I misunderstood your concern

Reply:
And agreed this is a problem, good catch!